test/integration/Ramm/swap.js

```diff
@@ -0,0 +1,224 @@
+    const deadline = nextBlockTimestamp + 5 * 60; // add 5 minutes
+
+    // Get expected book value
+    const initState = await ra.loadState();
+    const capital = await p1.getPoolValueInEth();
+    const supply = await tc.totalSupply();
+    const mcrValue = await mcr.getMCR();
+    const state = await ra._getReserves(initState, capital, supply, mcrValue, nextBlockTimestamp);
+    const k = state.eth.mul(state.nxmA);
+    const eth = state.eth.add(ethIn);
+    const nxmA = k.div(eth);
+    const expectedNxmOut = state.nxmA.sub(nxmA);
+
+    // Insufficient Amount Out Error
+    const minNxmOutFail = expectedNxmOut.add(1);
+    await setNextBlockBaseFee(0);
+    await setNextBlockTime(nextBlockTimestamp + 2 * 60);
+    const swapFail = ra.connect(member).swap(0, minNxmOutFail, deadline, { value: ethIn, maxPriorityFeePerGas: 0 });
+    await expect(swapFail).to.be.revertedWithCustomError(ra, 'InsufficientAmountOut');
+
+    // Minimum Amount Out Success
+    const minNxmOutSuccess = expectedNxmOut;
+    const before = await getCapitalSupplyAndBalances(p1, tc, tk, member.address);
+
+    await setNextBlockBaseFee(0);
+    await setNextBlockTime(nextBlockTimestamp + 3 * 60);
+    await ra.connect(member).swap(0, minNxmOutSuccess, deadline, { value: ethIn, maxPriorityFeePerGas: 0 });
+
+    const after = await getCapitalSupplyAndBalances(p1, tc, tk, member.address);
+    const nxmReceived = after.nxmBalance.sub(before.nxmBalance);
+    const transferFilter = tk.filters.Transfer(ethers.constants.AddressZero, member.address);
+    const nxmTransferEvents = await tk.queryFilter(transferFilter, -1);
+    const nxmTransferAmount = nxmTransferEvents[0]?.args?.value;
+
+    expect(after.ethCapital).to.be.equal(before.ethCapital.add(ethIn)); // ETH goes into capital pool
+    expect(after.nxmSupply).to.be.equal(before.nxmSupply.add(nxmReceived)); // NXM out is minted
+    expect(after.ethBalance).to.be.equal(before.ethBalance.sub(ethIn)); // member sends ETH
+    expect(after.nxmBalance).to.be.equal(before.nxmBalance.add(nxmTransferAmount)); // member receives NXM
+  });
+
+  it('should revert if ethOut < minAmountOut when swapping NXM for ETH', async function () {
+    const fixture = await loadFixture(setup);
+    const { ra, p1, tc, tk, mcr } = fixture.contracts;
+    const [member] = fixture.accounts.members;
+
+    const nxmIn = parseEther('1');
+
+    const { timestamp } = await ethers.provider.getBlock('latest');
+    // +3 hours to reach below BV, this ensures price is stable between the 2 swaps below
+    const nextBlockTimestamp = timestamp + 3 * 60 * 60;
+    const deadline = nextBlockTimestamp + 5 * 60; // add 5 minutes
+
+    // Get expected book value
+    const initState = await ra.loadState();
+    const capital = await p1.getPoolValueInEth();
+    const supply = await tc.totalSupply();
+    const mcrValue = await mcr.getMCR();
+    const state = await ra._getReserves(initState, capital, supply, mcrValue, nextBlockTimestamp);
```

Comment:
wasted a lot of time trying to get `ra.getReserves()` to work.

getReserves always has a different `block.timestamp` to what I set in `setNextBlockTime`, not really sure why.

so i ended up using the dirtier way of `ra._getReserves`.